hubspot/api.py

```diff
@@ -82,17 +100,44 @@ def get_sync_errors(limit=200, offset=0):
         query_params={"limit": limit, "offset": offset},
     )
     response.raise_for_status()
-    return response
+    return response.json().get("results", [])
+
+
+def get_sync_errors(limit=200, offset=0):
+    """
+    Yield hubspot errors
+
+    Args:
+        limit (int): The number of errors to be returned
+        offset (int): The index of the first error to be returned
+
+    Yields:
+        dict : error in JSON format
+    """
+    idx = 0
+    errors = paged_sync_errors(limit, offset)
+    num_errors = len(errors)
+    while idx < num_errors and num_errors:
+        for error in errors:
+            yield error
+            idx += 1
+            if idx == limit:
+                offset += limit
+                errors = paged_sync_errors(limit, offset)
+                num_errors = len(errors)
+                idx = 0
```

Comment:
I'm pretty sure this can be rewritten as:

```python
def get_sync_errors(limit=200, offset=0):
    errors = paged_sync_errors(limit, offset)
    while len(errors) > 0:
        yield from errors
        offset += limit
        errors = paged_sync_errors(limit, offset)
```

Reply:
That ends up calling `paged_sync_errors` far more often than necessary, once for each error found, for example if I add `print` statements:
```
response obtained with offset 0
response obtained with offset 1
response obtained with offset 2
response obtained with offset 3
response obtained with offset 4
response obtained with offset 5
response obtained with offset 6
response obtained with offset 7
```

Reply:
Modified it a bit this way:
```python
    errors = paged_sync_errors(limit, offset)
    while errors:
        for idx, error in enumerate(errors):
            yield error
            if idx == limit - 1:
                offset += limit
                errors = paged_sync_errors(limit, offset)
            elif idx == len(errors) - 1:
                errors = []
                break
```

Reply:
@mbertrand I must be missing something. I ran this standalone code sample with some canned return values for `paged_sync_errors`:

```python
def paged_sync_errors(limit=200, offset=0):
    print('paged_sync_errors')
    if offset == 0:
        return [1,2,3,4,5,6]
    if offset == 200:
        return [7,8,9,10]
    if offset == 400:
        return [11,12]
    return []

def get_sync_errors(limit=200, offset=0):
    errors = paged_sync_errors(limit, offset)
    while len(errors) > 0:
        yield from errors
        offset += limit
        errors = paged_sync_errors(limit, offset)
        
list(get_sync_errors())
```

Output:
```
paged_sync_errors
paged_sync_errors
paged_sync_errors
paged_sync_errors

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12] # <-- Result of list(get_sync_errors())
```

`paged_sync_errors` is called 4 times, and there are 12 items yielded by the generator. Not seeing this behavior where its called for every single error

Reply:
You're right, not sure what I did last Friday, but this still isn't working as originally implemented: it is now retrieving all errors in hubspot even if they occurred before the last error check.  Redid it this way:
```python
def get_sync_errors(timestamp, limit=200, offset=0):
    errors = paged_sync_errors(limit, offset)
    caught_up = False
    while len(errors) > 0 and not caught_up:
        for error in errors:
            if error.get("errorTimestamp") > timestamp:
                yield error
            else:
                caught_up = True
                break
        if not caught_up:
            offset += limit
            errors = paged_sync_errors(limit, offset)
```

Reply:
@mbertrand I'm assuming that the errors are returned in descending date order, correct? If that's the case, then yes, my generator above will return all errors if you keep iterating through it, but the consumer/client of that generator (`check_hubspot_api_errors` in this case) can `break` when it finds an error previous to the given timestamp. I think that's a much better approach than making the generator worry about the timestamp comparison

Reply:
Sorry, guess I need a vacation.  Changed.

Reply:
😆 no worries!!